playgrounds/react/src/examples/form/components/form.tsx

```diff
@@ -0,0 +1,73 @@
+import {
+  FormEventHandler, useMemo, useRef, useState,
+} from 'react';
+import { FormContext } from './form-context';
+import {
+  FormItemValidator, FormProps, FormValidationResult, Rule,
+} from './types';
+
+export function Form({ children, onSubmit }: FormProps) {
+  const [formValidationResult, setFormValidationResult] = useState<FormValidationResult>({});
+  const formValues = useRef<Record<string, unknown>>({});
+  const validationRules = useRef<Record<string, Rule[]>>({});
+  const validateFormItemValue: FormItemValidator = (value, rules) => {
+    const result: string[] = [];
+    if (rules?.length) {
+      rules.forEach((rule) => {
+        if (!rule.validate(value)) {
+          result.push(rule.message);
+        }
+      });
+    }
+    return result;
+  };
```

Comment:
Minor suggestion.

Let's make it shorter:
```suggestion
  const validateFormItemValue: FormItemValidator = (
    value, rules,
  ) => rules?.filter(({ validate }) => !validate(value)).map(({ message }) => message);
```

Reply:
Good point. Fixed.